### PR TITLE
[rl] add perf run config

### PR DESCRIPTION
**Pull request by wwwjn**

### CODE REVIEW

File: torchtitan/experiments/rl/examples/search_r1/config_registry.py [hunk 1]
```diff
@@ -136,3 +141,82 @@ def rl_grpo_qwen3_8b_search_r1() -> RLTrainer.Config:
...
+
+
+def rl_grpo_qwen3_30b_a3b_hybridep_search_r1() -> RLTrainer.Config:
+    """GRPO Search-R1 for Qwen3-30B-A3B MoE with a HybridEP + FULL-cudagraph generator.
+
+    Qwen3-30B-A3B has 4 KV heads, so the generator TP must be <=4. Applies the
+    ``fused_swiglu`` + ``helion_rope`` perf overrides (CUDA-only)
+    """
+    model_spec = model_registry(
+        "30B-A3B", attn_backend="varlen", moe_comm_backend="hybridep"
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/examples/search_r1/config_registry.py]: shall we change this to deepep to test multinode generator (say DP4TP4 EP16) with cudagraph on

Comment (tianyu-l) [reply torchtitan/experiments/rl/examples/search_r1/config_registry.py]: btw I think we should have two speed run
- Intranode only to compare with slime, on devgpu. Not sure if 30B model can fit.
- Internode maybe with 235B now that you could do EP16 with cudagraph (just realize capacity factoring tuning may make EP8 work)

File: torchtitan/experiments/rl/examples/search_r1/config_registry.py [hunk 2]
```diff
@@ -136,3 +141,82 @@ def rl_grpo_qwen3_8b_search_r1() -> RLTrainer.Config:
         ),
     )
     return config
+
+
+def rl_grpo_qwen3_30b_a3b_hybridep_search_r1() -> RLTrainer.Config:
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/examples/search_r1/config_registry.py]: add something to name that this is for speed run

File: torchtitan/experiments/rl/examples/search_r1/config_registry.py [hunk 3]
```diff
@@ -136,3 +141,82 @@ def rl_grpo_qwen3_8b_search_r1() -> RLTrainer.Config:
...
+    model_spec = model_registry(
+        "30B-A3B", attn_backend="varlen", moe_comm_backend="hybridep"
+    )
+    # qwen3's model_registry does not expose non_blocking_capacity_factor (deepseek_v3
+    # does), so set it per MoE layer post-build: a fixed capacity makes the HybridEP
+    # all-to-all static-shape and thus cudagraph-capturable under FULL.
+    for layer in model_spec.model.layers:
+        moe = getattr(layer, "moe", None)
+        if moe is not None:
+            moe.experts.token_dispatcher.non_blocking_capacity_factor = 1.0
```
Comment (wwwjn) [inline torchtitan/experiments/rl/examples/search_r1/config_registry.py]: 1.0 means worst case. 235B HybridEP might be droppable. If not well-load balanced, we might drop a lot of tokens. 

Qwen3 pre-training is using loss based, we only support loss-free.

File: torchtitan/experiments/rl/examples/search_r1/config_registry.py (hunk 1 above)
Comment (wwwjn) [reply torchtitan/experiments/rl/examples/search_r1/config_registry.py]: >  (just realize capacity factoring tuning may make EP8 work)

It's not possible to run 235B + HybridEP in 8GPU now, if failed when loading the weights from HF, not even hit the HybridEP buffer creation (the buffer is created lazily during first forward)

File: torchtitan/experiments/rl/examples/search_r1/config_registry.py [hunk 4]
```diff
@@ -141,3 +146,165 @@ def rl_grpo_qwen3_8b_search_r1() -> Controller.Config:
...
+        "torchtitan.overrides.helion_rope",
+    ]
+
+    return Controller.Config(
+        model_spec=model_spec,
+        hf_assets_path="torchtitan/experiments/rl/example_checkpoint/Qwen3-30B-A3B",
+        num_generators=2,  # TODO: TBD -- number of generator proc meshes to spawn
+        async_loop=AsyncLoopConfig(
+            num_training_steps=500,
+            num_groups_per_train_step=32,  # TODO: TBD
```
Comment (wwwjn) [inline torchtitan/experiments/rl/examples/search_r1/config_registry.py:183]: @felipemello1  I added a bunch of TBD here, maybe we can consolidate these 2 configs and discuss these fields

File: torchtitan/experiments/rl/examples/search_r1/config_registry.py (hunk 1 above)
Comment (tianyu-l) [reply torchtitan/experiments/rl/examples/search_r1/config_registry.py]: what's the error msg? btw I feel we should all-in DeepEP (inter / intra)

**Comment by felipemello1** [comment]
tuning params: https://docs.google.com/document/d/1u0jtpCtIvQsfTozmCQziyUMScMLlmG3XVGTtR1jEJq8/edit?tab=t.0

**Comment by yichuan-w** [comment]
one comment I have if that we might still need somewhere to have validation test, since we want to get the accuracy on testset, which seems deleted by @felipemello1 , but worth adding back somewhere or we can have a seperate node do eval

**Comment by felipemello1** [comment]
> one comment I have if that we might still need somewhere to have validation test, since we want to get the accuracy on testset, which seems deleted by @felipemello1 , but worth adding back somewhere or we can have a seperate node do eval

let me think about how we fit validation into the async loop

File: torchtitan/experiments/rl/examples/search_r1/config_registry.py [hunk 5]
```diff
@@ -145,3 +150,108 @@ def rl_grpo_qwen3_8b_search_r1() -> Controller.Config:
...
+            advantage=AdvantageEstimator.Config(should_std_normalize=True),
+        ),
+        renderer=RendererConfig(name="qwen3", enable_thinking=False),  # TODO: TBD
+        metrics=MetricsProcessor.Config(enable_wandb=True),
+        trainer=PolicyTrainer.Config(
+            optimizer=default_adamw(lr=1e-6),
+            lr_scheduler=LRSchedulersContainer.Config(
+                warmup_steps=2, decay_type="linear", min_lr_factor=1.0
+            ),
+            training=TrainingConfig(dtype="bfloat16"),
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/examples/search_r1/config_registry.py]: delete

File: torchtitan/experiments/rl/examples/search_r1/config_registry.py [hunk 6]
```diff
@@ -145,3 +150,108 @@ def rl_grpo_qwen3_8b_search_r1() -> Controller.Config:
...
+
+    config = Controller.Config(
+        model_spec=model_spec,
+        hf_assets_path="torchtitan/experiments/rl/example_checkpoint/Qwen3-30B-A3B",
+        num_generators=2,  # TODO: TBD -- number of generator proc meshes to spawn
+        async_loop=AsyncLoopConfig(
+            num_training_steps=500,
+            num_groups_per_train_step=32,  # TODO: TBD
+            group_size=8,  # TODO: TBD
+            validation=ValidationConfig(num_samples=500),
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/examples/search_r1/config_registry.py:185]: Is 500 common?

**Review by tianyu-l — APPROVED** [review]
trainer should use mixed precision with fp32 master weights

File: torchtitan/experiments/rl/examples/search_r1/config_registry.py [hunk 7]
```diff
@@ -199,7 +199,7 @@ def rl_grpo_qwen3_30b_a3b_deepep_search_r1_perf() -> Controller.Config:
             lr_scheduler=LRSchedulersContainer.Config(
                 warmup_steps=2, decay_type="linear", min_lr_factor=1.0
             ),
-            training=TrainingConfig(dtype="bfloat16"),
+            training=TrainingConfig(),
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/examples/search_r1/config_registry.py]: remove this line will give us default

**Review by tianyu-l — APPROVED** [review]
